feat (knn/kde) density estimation

geosarr opened the pull request:
# Multivariate density estimation

This PR implements density estimation from a sample using kd trees and nearest neighbors approach.

Bandwith is optionally provided by the caller.

geosarr committed:
feat: add knn density estimation with kdtree rs tree.

geosarr committed:
feat: add knn density estimation with kdtree rs tree.

geosarr committed:
feat: add a more generic implementation of knn density estimation.

geosarr committed:
feat: add a more generic implementation of knn density estimation.

geosarr committed:
perf: add performance bench.

geosarr committed:
perf: add performance bench.

geosarr committed:
feat: add bandwith support for knn density estimation.

geosarr committed:
feat: add bandwith support for knn density estimation.

geosarr committed:
feat: add first implementation of kernel density estimation.

geosarr committed:
feat: add first implementation of kernel density estimation.

geosarr committed:
fix: kernel density computation.

geosarr committed:
fix: kernel density computation.

geosarr committed:
perf: add performance bench.

geosarr committed:
perf: add performance bench.

geosarr committed:
chore: some refactoring.

geosarr committed:
chore: some refactoring.

geosarr committed:
chore: density functions borrowing evaluation pointS instead of owning them.

geosarr committed:
chore: density functions borrowing evaluation pointS instead of owning them.

geosarr committed:
fix imports and add some tests.

geosarr committed:
fix imports and add some tests.

YeungOnion commented:
This looks good, thanks for implementing!
As for knn wrapper on the tree and errors this seems good by me.

My only question is if you have some ideas on how you'd coordinate this with the Kernel trait to allow changing the kernel function from Gaussian?

geosarr committed:
add use of one dimensional gaussian kernel function.

geosarr committed:
add use of one dimensional gaussian kernel function.

YeungOnion committed:
fix: update to rand 0.10 and use seeded rng for tests

YeungOnion committed:
feat: gate density estimation behind a "kde" feature

density supports multivariate via nalgebra - so we're gating it
also adds thiserror, we'll eventually get to using it

YeungOnion committed:
docs: document DensityError variants and mark non_exhaustive

since thiserror was added for error propagation, figured we could use it
more directly.

YeungOnion committed:
doc: add paper reference and simple usage docs

YeungOnion committed:
docs: module docs and feature doc

YeungOnion committed:
chore: update lock MSRV for kde deps change
